src/cpuid/src/normalize.rs

```diff
@@ -0,0 +1,151 @@
+    /// Failed to set `CLFLUSH line size`.
+    #[error("Failed to set `CLFLUSH line size`: {0}")]
+    Clflush(CheckedAssignError),
+    /// Failed to get max CPUs per package.
+    #[error("Failed to get max CPUs per package: {0}")]
+    GetMaxCpusPerPackage(GetMaxCpusPerPackageError),
+    /// Failed to set max CPUs per package.
+    #[error("Failed to set max CPUs per package: {0}")]
+    SetMaxCpusPerPackage(CheckedAssignError),
+}
+
+/// Error type for `get_max_cpus_per_package`.
+#[derive(Debug, thiserror::Error, Eq, PartialEq)]
+pub enum GetMaxCpusPerPackageError {
+    /// Failed to get max CPUs per package as `cpu_count == 0`.
+    #[error("Failed to get max CPUs per package as `cpu_count == 0`")]
+    Underflow,
+    /// Failed to get max CPUs per package as `cpu_count > 128`.
+    #[error("Failed to get max CPUs per package as `cpu_count > 128`")]
+    Overflow,
+}
+
+// We use this 2nd implementation so we can conveniently define functions only used within
+// `normalize`.
+#[allow(clippy::multiple_inherent_impl)]
+impl super::Cpuid {
+    /// Applies required modifications to CPUID respective of a vCPU.
+    ///
+    /// # Errors
+    ///
+    /// When:
+    /// - [`IntelCpuid::normalize`] errors.
+    /// - [`AmdCpuid::normalize`] errors.
+    // As we pass through host frequency, we require CPUID and thus `cfg(cpuid)`.
+    #[inline]
+    pub fn normalize(
+        &mut self,
+        // The index of the current logical CPU in the range [0..cpu_count].
+        cpu_index: u8,
+        // The total number of logical CPUs.
+        cpu_count: u8,
+        // The number of bits needed to enumerate logical CPUs per core.
+        cpu_bits: u8,
+    ) -> Result<(), NormalizeCpuidError> {
+        // Update feature information entry
+        {
+            /// Flush a cache line size.
+            const EBX_CLFLUSH_CACHELINE: u32 = 8;
+
+            /// CPU is running on a hypervisor.
+            pub const HYPERVISOR_BITINDEX: u8 = 31;
+
+            /// The maximum number of logical processors per package is computed as the closest
+            /// power of 2 higher or equal to the CPU count configured by the user.
+            const fn get_max_cpus_per_package(
+                cpu_count: u8,
+            ) -> Result<u8, GetMaxCpusPerPackageError> {
+                // This match is better than but approximately equivalent to
+                // `2.pow((cpu_count as f32).log2().ceil() as u8)` (`2^ceil(log_2(c))`).
+                match cpu_count {
+                    0 => Err(GetMaxCpusPerPackageError::Underflow),
+                    1 => Ok(1),
+                    2 => Ok(2),
+                    3..=4 => Ok(4),
+                    5..=8 => Ok(8),
+                    9..=16 => Ok(16),
+                    17..=32 => Ok(32),
+                    33..=64 => Ok(64),
+                    65..=128 => Ok(128),
+                    129..=u8::MAX => Err(GetMaxCpusPerPackageError::Overflow),
+                }
```

Comment:
Can we use [`u8::checked_next_power_of_two`](https://doc.rust-lang.org/std/primitive.u8.html#method.checked_next_power_of_two) here?

Reply:
We could, but it adds an extra step e.g.
```rust
match cpu_count {
    0 => Err(GetMaxCpusPerPackageError::Underflow),
    x => u8::checked_next_power_of_two(x).ok_or(GetMaxCpusPerPackageError::Overflow)
}

```